// react-static-generator/lib/snapshots.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _jsdom = require('jsdom');

var _jsdom2 = _interopRequireDefault(_jsdom);

var _writer = require('./writer');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var snap = function snap(_ref) {
	var baseUrl = _ref.baseUrl,
	    baseDir = _ref.baseDir;

	var _url$parse = _url2.default.parse(baseUrl),
	    protocol = _url$parse.protocol,
	    host = _url$parse.host,
	    path = _url$parse.path;

	return new Promise(function (resolve, reject) {
		_jsdom2.default.env({
			url: protocol + '//' + host + path,
			resourceLoader: function resourceLoader(resource, callback) {
				if (resource.url.host === host) {
					resource.defaultFetch(callback);
				} else {
					callback();
				}
			},

			features: {
				FetchExternalResources: ["script"],
				ProcessExternalResources: ["script"],
				SkipExternalResources: false
			},
			virtualConsole: _jsdom2.default.createVirtualConsole().sendTo(console),
			done: function done(err, window) {
				if (err) reject(err);
				resolve(window);
			}
		});
	}).then(function (window) {
		var content = _jsdom2.default.serializeDocument(window.document);
		var filename = void 0;
		if (path === "/") {
			filename = 'index.html';
		} else {
			filename = path + '/index.html';
		}
		console.log('\u270F\uFE0F   Saving ' + path + ' as ' + filename);
		return (0, _writer.write)({ baseDir: baseDir, filename: filename, content: content });
	}).catch(function (err) {
		console.log("error: ", error);
	});
};

// return array of promises

exports.default = function (_ref2) {
	var baseDir = _ref2.baseDir,
	    paths = _ref2.paths,
	    port = _ref2.port;

	var snapshots = paths.map(function (path) {
		return snap({ baseUrl: 'http://localhost:' + port + path, baseDir: baseDir });
	});
	return Promise.all(snapshots);
};

module.exports = exports['default'];
